File: src/main/java/org/edtp/universe/player/HumanPlayerDetector.java
package org.edtp.universe.player;

import net.minecraft.network.Connection;
import net.minecraft.server.level.ServerPlayer;
import org.edtp.universe.mixin.ServerCommonPacketListenerAccessor;

public final class HumanPlayerDetector {
    private static final String CARPET_FAKE_PLAYER = "carpet.patches.EntityPlayerMPFake";

    private HumanPlayerDetector() {
    }

    public static boolean isHuman(ServerPlayer player) {
        if (isCarpetFake(player)) {
            return false;
        }
        Connection network = ((ServerCommonPacketListenerAccessor) player.connection)
            .universe647$getConnection();
        return network.isConnected();
    }

    private static boolean isCarpetFake(ServerPlayer player) {
        Class<?> type = player.getClass();
        while (type != null) {
            if (type.getName().equals(CARPET_FAKE_PLAYER)) {
                return true;
            }
            type = type.getSuperclass();
        }
        return false;
    }
}

File: src/main/kotlin/org/edtp/universe/player/HumanPlayerDetector.kt
package org.edtp.universe.player

import net.minecraft.server.level.ServerPlayer
import org.edtp.universe.mixin.ServerCommonPacketListenerAccessor

object HumanPlayerDetector {
    private const val CARPET_FAKE_PLAYER = "carpet.patches.EntityPlayerMPFake"

    fun isHuman(player: ServerPlayer): Boolean {
        if (isCarpetFake(player)) {
            return false
        }
        val network = (player.connection as ServerCommonPacketListenerAccessor).`universe647$getConnection`()
        return network.isConnected
    }

    private fun isCarpetFake(player: ServerPlayer): Boolean {
        var type: Class<*>? = player.javaClass
        while (type != null) {
            if (type.name == CARPET_FAKE_PLAYER) {
                return true
            }
            type = type.superclass
        }
        return false
    }
}
